### DISCUSSION TITLE
How to explicitly encode no key signature

### DISCUSSION THREAD

**Discussion by rettinghaus** [music-encoding/music-encoding#967] [Q&A]
Atonal music usually has no written key, but that is not the same as `<keySig sig="0">`, which means C major. Visually there is no difference, but logically there is a big difference. If you transpose an atonal piece up one tone, in this case the key would change to D major and two sharps would appear `(<keySig sig="2s">)`.
Not encoding a `keySig`, on the other hand, can mean that "there are simply no written accidentals" and could thus imply C major. 
SibMEI does the former, which I would think is wrong. But then is the latter the recommended _best practice_ of encoding?

Or would we need a way to make "no key signature" more explicit by adding `"none"` or `"atonal"` values to `data.KEYFIFTHS`?

**Comment by craigsapp** [comment]
Lack of key signature is also a property of timpani, and often in brass instruments.

Here is an example:

<img width="709" alt="Screen Shot 2022-05-23 at 6 31 27 AM" src="https://user-images.githubusercontent.com/3487289/169830683-ace859bd-11e0-49df-9059-99be463d0042.png">

https://s9.imslp.org/files/imglnks/usimg/b/b9/IMSLP06803-Beethoven_-_Symphony_No.7_Mvt.IV_(ed._Unger).pdf

The music is in A major, so three sharps are expected in the key signature (such as in the violins).   Notice there are four instruments that do not have a key signature: the clarinets, French horns, trumpets and timpani.

The clarinets are in A, so they do not have any sharps/flats in their key signature since they are playing in C major from their point of view.

The timpani does not have any accidentals in the key signature since they don't play many notes.  In this case their key signature is not C major (or A minor) with a fifths value of 0, but literally no key signature, with a fifths value of undefined.

The two brass instruments are interesting.  Even though the French horns (Corni in Italian) are in A and you would expect the key signature to be C major, they actually have no key signature in the same manner as the timpani.   If you were to transpose the symphony but keep using horns in A, the horn parts should still have no key signature.

The trumpet parts are an interesting case that demonstrate that there is no key signature for the brass parts in this score:  They are in D, so you would expect the key signature to have one sharp because they are playing in G major from their point of view.

--------------------------------------

In an encoding of this score, I believe that the brass and timpani parts should be encoded without a `<keySig>` (or at least without information about the accidentals in the key signature such as the `keySig@sig` attribute).   When you transpose the music, there would be no key signature to also translate.   This is unlike `<keySig sig="0">` which should be transposed along with the music.  

This would be similar to atonal music.  Looking at Webern's Six pieces for orchestra, op. 6, clarinet part, first edition parts and score:

<img width="754" alt="Screen Shot 2022-05-23 at 8 04 33 AM" src="https://user-images.githubusercontent.com/3487289/169849461-a38028d9-0534-4a23-badb-208a31593bb2.png">

https://imslp.eu/files/imglnks/euimg/4/4a/IMSLP578085-PMLP2079-3_cl.pdf

The first note is an F5 written, which is an E-flat sounding.

Here is the clarinet part in the full score:

<img width="905" alt="Screen Shot 2022-05-23 at 8 08 13 AM" src="https://user-images.githubusercontent.com/3487289/169850179-591ff87c-7b10-4323-a4b9-d413b8f395e3.png">

https://s9.imslp.org/files/imglnks/usimg/4/44/IMSLP440065-PMLP02079-webernsechsstuckeop.61909versionfullscore.pdf

The part (first example) is clearly at written pitch since there would be no purpose in giving a B-flat clarinet a sounding part.   Looking at the score, it is clear that this is a written score since the pitches of the clarinet are the same as in the part.  

But there is a key signature in the score, while there is none in the part.  The two-sharp key signature in the score is probably used to indicate that the part is transposed (and the conductor should be able to decode the sounding pitch faster by having the key signature present in the full score.   For the clarinet player, the key signature would be a hinderance and unnecessary since all of their notes have explicit accidentals.

I would say that the clarinet part has an undefined key signature, while the clarinet staff in the score has a key signature.

In this edition of the score:

<img width="931" alt="Screen Shot 2022-05-23 at 8 19 42 AM" src="https://user-images.githubusercontent.com/3487289/169852895-c2bb85c5-968a-4703-beb0-9f5218e9dcd9.png">
https://imslp.eu/files/imglnks/euimg/2/2d/IMSLP424298-PMLP02079-Op._6_6_Pieces_Original_copy.pdf

They got rid of the key signatures in the full score (which would make me confused since it looks like a sounding score because all of the instruments have the same key signature &mdash; you would have to compare the score to the parts to verify that the score is in transposed pitch).




---------------------------------------------

Verovio's transposition capabilities currently cannot handle the case of an undefined key signature properly.    When transposing music with a key signature, notes with `@accid` keep `@accid` and notes with `@accid.ges` keep `@accid.ges` and there is no switching between these two parameters (i.e., a visual accidental remains a visual accidental and an implied accidental remains an implied accidental).

However, with parts that do not have a key signature, there will be changes from `@acc` to `@acc.ges` as well as `@acc.ges` to `@acc`.

An exception is that atonal music with a visual accidental on every note will be transposed properly by verovio currently since all accidentals use `@accid` (such as the clarinet part in the Webern example above).

For an atonal key signature, I would say to either leave out `<keySig>` completely, or otherwise do not provide `keySig@sig`, and the atonalness would be encoded in `@mode` such as:   `<keySig mode="atonal">`.


--------------------------

> Not encoding a keySig, on the other hand, can mean that "there are simply no written accidentals" and could thus imply C major.

I would say that there is no implication of C major in such cases.   For C major you must encode a key signature and include the mode:

```xml
<keySig sig="0" mode="major" pname="c">
```

The reason why C major cannot be implied, is that this is also the key signature of A minor:

```xml
<keySig sig="0" mode="minor" pname="a">
```

Not to mention modal music:

```xml
<keySig sig="0" mode="locrian" pname="b">
```

So there can be no implication when you have a missing `@mode` (`@pname` can be derived from `@sig` and `@mode`, however).

By extension,  atonal music should be:

```
<keySig mode="atonal">
```

where there is no `@sig` and no `@pname`.


Also note that the music can modulate to a new key, but keep the same key signature.   This is common in sonata-allegro forms, for example, since the second theme is usually in the dominant key, but the key signature usually does not change since the key is close to the tonic key (adding only one sharp).   So I would view this key signature as valid:

```
<keySig sig="0" mode="major" pname="g">
```

This would be music in G major that does not have sharps or flats in the key signature (so there would be written F-sharps in the score).   For analytic purposes, it would be useful to encode the key information without the key signature:

```
<keySig mode="major" pname="g">
```

This will not produce a logical key signature, but instead indicates what the key of the music is after the modulation.

Humdrum decouples key and key signature:

```tsv
**kern
*k[]
*G:
=1
4g
4a
4b
4cc
=2
4dd
4ee
4ff#
4gg
=
*-
```

This gets notated with verovio as:

<img width="627" alt="Screen Shot 2022-05-23 at 8 54 37 AM" src="https://user-images.githubusercontent.com/3487289/169859302-48e7b949-52d0-4bfe-af2a-3aaa60d0263e.png">

<details>
<summary>
Click for full MEI example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-23T08:54:20" version="3.10.0-dev-1b4035b">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mrjacep">
    <score xml:id="s9j6rcl">
     <scoreDef xml:id="swxnvev">
      <staffGrp xml:id="s7c4ubr">
       <staffDef xml:id="staffdef-L1F1" n="1" lines="5">
        <clef xml:id="cq63jkg" shape="G" line="2" />
        <keySig xml:id="keysig-L2F1" pname="g" mode="major" sig="0" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="staff-L1F1" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="4" oct="4" pname="g" accid.ges="n" />
         <note xml:id="note-L6F1" dur="4" oct="4" pname="a" accid.ges="n" />
         <note xml:id="note-L7F1" dur="4" oct="4" pname="b" accid.ges="n" />
         <note xml:id="note-L8F1" dur="4" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
      </measure>
      <measure xml:id="measure-L9" n="2">
       <staff xml:id="staff-L9F1N1" n="1">
        <layer xml:id="layer-L9F1N1" n="1">
         <note xml:id="note-L10F1" dur="4" oct="5" pname="d" accid.ges="n" />
         <note xml:id="note-L11F1" dur="4" oct="5" pname="e" accid.ges="n" />
         <note xml:id="note-L12F1" dur="4" oct="5" pname="f" accid="s" />
         <note xml:id="note-L13F1" dur="4" oct="5" pname="g" accid.ges="n" />
        </layer>
       </staff>
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

where the keySig merges the "key designation" and the "key signature" information into a single element:

```xml
<keySig xml:id="keysig-L2F1" pname="g" mode="major" sig="0" />
```
